### ISSUE TITLE
rdkafka2 - Can't set partition_key

### ISSUE THREAD

**Issue by kevinkirkup** [fluent/fluent-plugin-kafka#264]
When attempting to use the `rdkafka2` output plugin, if I just set the `default_partition_key` the same way I do for the `kafka2` plugin, it completely ignores it and send to a partition based on the message content.

If I set the `partition_key` to a field to use for calculating the partition, I get the following exception:
```
2019-05-28 21:40:21 +0000 [debug]: #0 Created new chunk chunk_id="589f983062baffa1e8f1359651c80a9a" metadata=#<struct Fluent::Plugin::Buffer::Metadata timekey=1559079620, tag=nil, variables=nil>
2019-05-28 21:40:28 +0000 [warn]: #0 Send exception occurred: undefined method `code' for #<TypeError: no implicit conversion of String into Integer> at /opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/fluent-plugin-kafka-0.9.4/lib/fluent/plugin/out_rdkafka2.rb:277:in `rescue in block in enqueue_with_retry'
2019-05-28 21:40:28 +0000 [warn]: #0 got unrecoverable error in primary and no secondary error_class=NoMethodError error="undefined method `code' for #<TypeError: no implicit conversion of String into Integer>"
  2019-05-28 21:40:28 +0000 [warn]: #0 /opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/fluent-plugin-kafka-0.9.4/lib/fluent/plugin/out_rdkafka2.rb:277:in `rescue in block in enqueue_with_retry'
  2019-05-28 21:40:28 +0000 [warn]: #0 /opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/fluent-plugin-kafka-0.9.4/lib/fluent/plugin/out_rdkafka2.rb:274:in `block in enqueue_with_retry'
  2019-05-28 21:40:28 +0000 [warn]: #0 /opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/fluent-plugin-kafka-0.9.4/lib/fluent/plugin/out_rdkafka2.rb:273:in `loop'
  2019-05-28 21:40:28 +0000 [warn]: #0 /opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/fluent-plugin-kafka-0.9.4/lib/fluent/plugin/out_rdkafka2.rb:273:in `enqueue_with_retry'
  2019-05-28 21:40:28 +0000 [warn]: #0 /opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/fluent-plugin-kafka-0.9.4/lib/fluent/plugin/out_rdkafka2.rb:259:in `block in write'
  2019-05-28 21:40:28 +0000 [warn]: #0 /opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/fluentd-1.4.2/lib/fluent/event.rb:323:in `each'
  2019-05-28 21:40:28 +0000 [warn]: #0 /opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/fluentd-1.4.2/lib/fluent/event.rb:323:in `block in each'
  2019-05-28 21:40:28 +0000 [warn]: #0 /opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/fluentd-1.4.2/lib/fluent/plugin/buffer/chunk.rb:195:in `block in open'
  2019-05-28 21:40:28 +0000 [warn]: #0 /opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/fluentd-1.4.2/lib/fluent/plugin/buffer/file_chunk.rb:172:in `open'
  2019-05-28 21:40:28 +0000 [warn]: #0 /opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/fluentd-1.4.2/lib/fluent/plugin/buffer/chunk.rb:187:in `open'
  2019-05-28 21:40:28 +0000 [warn]: #0 /opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/fluentd-1.4.2/lib/fluent/event.rb:322:in `each'
  2019-05-28 21:40:28 +0000 [warn]: #0 /opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/fluent-plugin-kafka-0.9.4/lib/fluent/plugin/out_rdkafka2.rb:241:in `write'
  2019-05-28 21:40:28 +0000 [warn]: #0 /opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/fluentd-1.4.2/lib/fluent/plugin/output.rb:1125:in `try_flush'
  2019-05-28 21:40:28 +0000 [warn]: #0 /opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/fluentd-1.4.2/lib/fluent/plugin/output.rb:1425:in `flush_thread_run'
  2019-05-28 21:40:28 +0000 [warn]: #0 /opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/fluentd-1.4.2/lib/fluent/plugin/output.rb:454:in `block (2 levels) in start'
  2019-05-28 21:40:28 +0000 [warn]: #0 /opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/fluentd-1.4.2/lib/fluent/plugin_helper/thread.rb:78:in `block in thread_create'
2019-05-28 21:40:28 +0000 [warn]: #0 bad chunk is moved to /tmp/fluent/backup/worker0/object_2b1524061a44/589f983062baffa1e8f1359651c80a9a.log
```

If I set a `partition` field to a fixed value, I get:
```
2019-05-28 21:42:33 +0000 [debug]: #0 Created new chunk chunk_id="589f98ae0b4832897be3e99239a65a53" metadata=#<struct Fluent::Plugin::Buffer::Metadata timekey=1559079750, tag=nil, variables=nil>
2019-05-28 21:42:37 +0000 [warn]: #0 Send exception occurred: undefined method `code' for #<TypeError: no implicit conversion of String into Integer> at /opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/fluent-plugin-kafka-0.9.4/lib/fluent/plugin/out_rdkafka2.rb:277:in `rescue in block in enqueue_with_retry'
2019-05-28 21:42:37 +0000 [warn]: #0 got unrecoverable error in primary and no secondary error_class=NoMethodError error="undefined method `code' for #<TypeError: no implicit conversion of String into Integer>"
  2019-05-28 21:42:37 +0000 [warn]: #0 /opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/fluent-plugin-kafka-0.9.4/lib/fluent/plugin/out_rdkafka2.rb:277:in `rescue in block in enqueue_with_retry'
  2019-05-28 21:42:37 +0000 [warn]: #0 /opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/fluent-plugin-kafka-0.9.4/lib/fluent/plugin/out_rdkafka2.rb:274:in `block in enqueue_with_retry'
  2019-05-28 21:42:37 +0000 [warn]: #0 /opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/fluent-plugin-kafka-0.9.4/lib/fluent/plugin/out_rdkafka2.rb:273:in `loop'
  2019-05-28 21:42:37 +0000 [warn]: #0 /opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/fluent-plugin-kafka-0.9.4/lib/fluent/plugin/out_rdkafka2.rb:273:in `enqueue_with_retry'
  2019-05-28 21:42:37 +0000 [warn]: #0 /opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/fluent-plugin-kafka-0.9.4/lib/fluent/plugin/out_rdkafka2.rb:259:in `block in write'
  2019-05-28 21:42:37 +0000 [warn]: #0 /opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/fluentd-1.4.2/lib/fluent/event.rb:323:in `each'
  2019-05-28 21:42:37 +0000 [warn]: #0 /opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/fluentd-1.4.2/lib/fluent/event.rb:323:in `block in each'
  2019-05-28 21:42:37 +0000 [warn]: #0 /opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/fluentd-1.4.2/lib/fluent/plugin/buffer/chunk.rb:195:in `block in open'
  2019-05-28 21:42:37 +0000 [warn]: #0 /opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/fluentd-1.4.2/lib/fluent/plugin/buffer/file_chunk.rb:172:in `open'
  2019-05-28 21:42:37 +0000 [warn]: #0 /opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/fluentd-1.4.2/lib/fluent/plugin/buffer/chunk.rb:187:in `open'
  2019-05-28 21:42:37 +0000 [warn]: #0 /opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/fluentd-1.4.2/lib/fluent/event.rb:322:in `each'
  2019-05-28 21:42:37 +0000 [warn]: #0 /opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/fluent-plugin-kafka-0.9.4/lib/fluent/plugin/out_rdkafka2.rb:241:in `write'
  2019-05-28 21:42:37 +0000 [warn]: #0 /opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/fluentd-1.4.2/lib/fluent/plugin/output.rb:1125:in `try_flush'
  2019-05-28 21:42:37 +0000 [warn]: #0 /opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/fluentd-1.4.2/lib/fluent/plugin/output.rb:1425:in `flush_thread_run'
  2019-05-28 21:42:37 +0000 [warn]: #0 /opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/fluentd-1.4.2/lib/fluent/plugin/output.rb:454:in `block (2 levels) in start'
  2019-05-28 21:42:37 +0000 [warn]: #0 /opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/fluentd-1.4.2/lib/fluent/plugin_helper/thread.rb:78:in `block in thread_create'
2019-05-28 21:42:37 +0000 [warn]: #0 bad chunk is moved to /tmp/fluent/backup/worker0/object_2add9a9d135c/589f98ae0b4832897be3e99239a65a53.log
^C2019-05-28 21:42:39 +0000 [debug]: #0 fluent/log.rb:302:debug: fluentd main process get SIGINT
```

**Comment by kevinkirkup** [comment]
Use 
```
root@virt-pod-17a-e2:/home/vagrant# td-agent --version
td-agent 1.4.2
```

APT Info
```
root@virt-pod-17a-e2:/home/vagrant# apt-cache policy td-agent
td-agent:
  Installed: 3.4.1-0
  Candidate: 3.4.1-0
  Version table:
 *** 3.4.1-0 500
        500 http://packages.treasuredata.com/3/ubuntu/bionic bionic/contrib amd64 Packages
        100 /var/lib/dpkg/status
     3.4.0-0 500
        500 http://packages.treasuredata.com/3/ubuntu/bionic bionic/contrib amd64 Packages
     3.3.0-1 500
        500 http://packages.treasuredata.com/3/ubuntu/bionic bionic/contrib amd64 Packages
     3.3.0-0 500
        500 http://packages.treasuredata.com/3/ubuntu/bionic bionic/contrib amd64 Packages
     3.2.1-0 500
        500 http://packages.treasuredata.com/3/ubuntu/bionic bionic/contrib amd64 Packages
     3.2.0-0 500
        500 http://packages.treasuredata.com/3/ubuntu/bionic bionic/contrib amd64 Packages
     2.5.0-0 500
        500 http://packages.treasuredata.com/2.5/ubuntu/bionic bionic/contrib amd64 Packages
```

**Comment by repeatedly** [comment]
The original error is TypeError so your specified `partition_key` value is wrong type. Maybe string?

BTW, following error is bad.

https://github.com/fluent/fluent-plugin-kafka/blob/fe1db0634f604a92d3cfb062bf37a1e0b8955e12/lib/fluent/plugin/out_rdkafka2.rb#L277

This point assumes error type has `code` attribute but some plane errors don't have it.
I will fix this.

**Comment by kevinkirkup** [comment]
@repeatedly For the `TypeError` above, here are the configurations that was using:

```
<filter app.frr>
  @type record_transformer
  enable_ruby

  <record>
    host_pod "{{ host_pod_id }}"
    container_name ${record["log_path"].split("/")[5]}
    nodetype {{ fluentd_nodetype }}
    host "#{Socket.gethostname}"
    subsystem {{ fluentd_subsystem }}
    component {{ fluentd_component }}
    partition 41                                                  <----- here
  </record>
</filter>
```

Here specifying the `partition_key`.
```
<match app.frr>
  @type kafka_buffered

  brokers {{ fluentd_kafka_brokers }}

  buffer_type file
  buffer_path /var/log/td-agent/buffer/frr
  flush_interval 1

  # topic/partition settings
  default_topic {{ fluentd_topics.frr_default }}
  partition_key container_name                   <---- here 
...
</match>
```
`container_name` field here is a string (e.g. "instance-xxxxxxx").
When using the `kafka_buffered` plugin, this string was hashed to calculate the partition.

**Comment by repeatedly** [comment]
I see.
ruby-kafka uses crc32 to calculate the partition: https://github.com/zendesk/ruby-kafka/blob/1b6ee76d16a75475f209bea9d684bbd06a262108/lib/kafka/partitioner.rb#L31
rdkafka-ruby paases parition value directly to rdkafka: https://github.com/appsignal/rdkafka-ruby/blob/87b3e0ddae5ea576847cb67228fcea06ec2a24d6/lib/rdkafka/producer.rb#L110
This is why this error happens.

So there are 2 main approaches:
- Use similar approach in rdkafka plugin
- Set number in record_transformer

**Comment by kenhys** [comment]
Fixed in fluent-plugin-kafka v0.10.0
